Encourage social media share

alejandronanez opened the pull request:
Add method to tell wether the logged in user is the author of the thread or not

<!-- FILL OUT THE BELOW FORM OR YOUR PR WILL BE AUTOMATICALLY CLOSED -->
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- hyperion (frontend)

----

@mxstbr I added ~`lodash.get`~ `idx` so we can start doing:

~FWIW: [_.get docs](https://lodash.com/docs/4.17.5#get)~

```javascript
const foo = idx(this, _ => _.props.data.a.really.deep.nested.prop);

// instead of

const foo = this.props &&
    this.props.data &&
    this.props.data.a &&
    this.props.data.a.really ... // and so on
```

~Let me know what do you think.~

### Not author's thread
![screen shot 2018-05-08 at 9 49 42 am](https://user-images.githubusercontent.com/464978/39764403-2be87422-52a5-11e8-8811-7da7500b49c9.png)


### Author's thread
![screen shot 2018-05-08 at 9 46 29 am](https://user-images.githubusercontent.com/464978/39764340-0e29a514-52a5-11e8-8bd3-9469f746e690.png)

<!--

If your pull request introduces changes to the user interface on Spectrum, please share before and after screenshots of the changes (gifs or videos are encouraged for interaction changes). Please include screenshots of desktop and mobile viewports to ensure that all responsive cases are reviewed.

-->

Closes #3040

alejandronanez committed:
Add lodash.get
Add method to tell wether the logged in user is the author of the thread or not

alejandronanez committed:
rolling back content in NullMessagesWrapper

mxstbr commented on src/views/thread/components/messages.js:
```diff
@@ -48,6 +49,7 @@ type Props = {
   data: {
     thread: {
```
Let's type this as `import type { ThreadInfoType } from 'shared/graphql/thread/threadInfo'; data: { thread: ThreadInfoType } }`

mxstbr reviewed (commented):
That is a very sensible suggestion! I'd prefer we use [`idx`](https://github.com/facebookincubator/idx) instead because it has 0 runtime overhead :blush: 

Other than that looks good to me so far

ryota-murakami commented:
I'm struggling annoying undefined check in JavaScript, it's so impressive:grinning:

alejandronanez commented:
> That is a very sensible suggestion! I'd prefer we use idx instead because it has 0 runtime overhead 😊

haha, np. I'll add `idx` instead of `lodash.get` 👍

alejandronanez committed:
Add social sharing buttons

alejandronanez commented on src/views/thread/style.js:
```diff
@@ -160,6 +163,85 @@ export const ThreadHeading = styled(H1)`
+  }
+`;
+
+export const SocialIcon = styled(Icon)``;
```
@mxstbr I couldn't make this styled-component work :( that's why I had to do 

```javascript
export const BaseButton = styled.div`
 .icon {
     opacity: 1 !important;
  }

  svg {
     fill: ${({ theme }) => theme.bg.default};
  }
`
```

Instead of:

```javascript
export const SocialIcon = styled(Icon)`
  opacity: 1; // Opacity didn't change and the class for this styled-component was not applied to the Icon element.
`;
```

Maybe is something related with the `Icon` component itself?

superbryntendo replied:
We shouldn't use the Icon component here - we should use the Button component from src/components/buttons and set the glyph prop to the icon you prefer.

alejandronanez replied:
Thanks for the heads up @uberbryn. 👍

brianlovin commented:
We already have a graphql type for `isAuthor` on the thread model - see `api/queries/thread/isAuthor` :)

brianlovin commented:
I must be missing something, because it looks like you're using that field already. Why is a new package needed here?

Sorry for my confusion!

brianlovin commented:
And re: styling - can we keep the same type as the empty state for messages? It should be smaller and a bit more subdued - happy to jump in and polish that up if needed.

alejandronanez commented:
> Why is a new package needed here?

np @brianlovin, we can use `idx` to access deeply nested properties with ease. Look this example, you can do:

```javascript
// By having this:
this.props = {
  data: {
    a: {
      really: {
        deep: {
          nested: {
            prop: 'brian'
          }
        }
      }
    }
  }
}

// You can doo:

const name = idx(this, _ => _.props.data.a.really.deep.nested.prop); // brian

// instead of

const name = this.props &&
    this.props.data &&
    this.props.data.a &&
    this.props.data.a.really ... // and so on
```

If any of those properties does not exist, you'll get `undefined` instead.

Hopefully, this makes sense.

You can read more about `idx` here: https://github.com/facebookincubator/idx

alejandronanez commented:
> I must be missing something, because it looks like you're using that field already.

I'm getting it from `this.props.data.thread.isAuthor`, that's correct, right?

alejandronanez commented:
> And re: styling - can we keep the same type as the empty state for messages? It should be smaller and a bit more subdued - happy to jump in and polish that up if needed.

Can you upload a mock please? If you're not rushed about this task I'll update the styles tomorrow or later today

brianlovin commented:
Ah, I see - the ability to return undefined at any point in the chain seems super useful, there are a ton of places we could use that to clean up the code!

I won't have time for a mock today, but can jump on this whenever everything else is ready to go and tidy it up :)

alejandronanez commented:
> Ah, I see - the ability to return undefined at any point in the chain seems super useful, there are a ton of places we could use that to clean up the code!

Yes!

> I won't have time for a mock today, but can jump on this whenever everything else is ready to go and tidy it up :)

If I finish it tomorrow (tests are still pending), do you think it will be _ok_ to merge it and iterate on it later? cc @brianlovin @mxstbr

brianlovin commented:
The UI will just need some polish before we merge it in - should just take a second when this is ready :)

alejandronanez committed:
Merge branch 'alpha' of github.com:withspectrum/spectrum into encourage-social-media-share-issue#3040

alejandronanez committed:
update type for thread

alejandronanez committed:
fix ThreadInfoType route

alejandronanez committed:
update flow types

alejandronanez committed:
fix flowType errors - finally

alejandronanez commented:
@brianlovin Ok, I completed my work on this, when you have the mocks ready I can come back and finish this task :)

I'll see what I can grab next in the meantime 👍

alejandronanez committed:
Merge branch 'alpha' of github.com:withspectrum/spectrum into encourage-social-media-share-issue#3040

alejandronanez committed:
update social buttons

alejandronanez committed:
- Remove FB and Twitter buttons from ./style.js

alejandronanez commented:
I just got rid of `FacebookButton` and `TwitterButton` components from `style.js` after knowing the existense of `<Button gradientTheme={'social.facebook'}` in [this pr](https://github.com/withspectrum/spectrum/pull/3082/files#diff-aefb8bed155fa774112ad531678c665aR82). Thanks @uberbryn !

mxstbr commented:
Sorry for the super slow work here, I'm picking this back up! Will merge conflicts, maybe adjust the styling a little bit and then shipit 🚢

mxstbr committed:
Merge branch 'alpha' into alejandronanez-encourage-social-media-share-issue#3040

mxstbr committed:
Fix flash of null state on first load of inbox

Due to the thread not having been loaded yet, it would first show the
old upsell for the non-author and only after a second or two would
switch to the author upsell. Kinda jarring.

This fixes it by waiting for the thread to load before showing the upsell!

mxstbr committed:
Fix icon opacity in upsell

mxstbr committed:
Change author empty thread upsell wording

mxstbr commented:
Adjusted the styling and copy a little bit:

![screen shot 2018-08-06 at 14 00 48](https://user-images.githubusercontent.com/7525670/43715591-464c6256-9981-11e8-9939-12a810fb8af4.png)

Now it fits well with the other upsell, looking similar:

![screen shot 2018-08-06 at 13 58 27](https://user-images.githubusercontent.com/7525670/43715599-4dd7321c-9981-11e8-8f2e-ba5f123784d2.png)

Pending a quick review by @brianlovin this is ready to go!

mxstbr committed:
Fix security hole in sharing

alejandronanez commented:
Yay!

I’m glad to see this moving forward! :-)

brianlovin commented:
@alejandronanez this is super close! Last thing is it just needs a tiny bit of padding on mobile, or perhaps the buttons should stack vertically on mobile so that the bbuttons don't run up against the left and right edges of the screen. Then let's ship and iterate!

mxstbr committed:
Fix mobile styling of thread share upsell

mxstbr reviewed (approved):
Donezo, took care of that really quickly! Shipping :ship:

brianlovin commented:
@mxstbr dumb workflow question, but when you commit directly there are you doing it on the fork? Or do you have a faster way to get involved in a contributor branch?

mxstbr commented:
Doing it on the fork! Using https://github.com/github/hub makes it a hell of a lot easier:

```sh
# Checkout the PR #3042 (this one)
hub pr checkout 3042
# ...work on stuff...
git commit -m 'Bla'
# Push to the fork
git push git@github.com:alejandronanez/spectrum.git alejandronanez-encourage-social-media-share-issue\#3040:encourage-social-media-share-issue#3040
```

That last `git push` is the hardest part: you have to push to their fork, but locally the branch is called `<username>-<branchname>`, where on the fork it's obviously only `branchname`, so we use the `:` delimiter to push our locally named branch onto a different branch in the fork. Not sure if this makes sense, but it's the fastest way I've found of doing this!
